File: src/python-app/src/infrastructure/file_reader.py
"""File reading infrastructure."""

from pathlib import Path
from typing import Protocol

from src.domain.exceptions import FileOperationException


class FileReader(Protocol):
    """Protocol for file reading operations."""
    
    def read_lines(self, path: str) -> list[str]:
        """Read lines from file."""
        ...


class TextFileReader:
    """
    Reads text files containing network records.
    
    Handles file validation and encoding issues.
    """
    
    def __init__(self, encoding: str = "utf-8"):
        """
        Initialize file reader.
        
        Args:
            encoding: Text encoding (default: utf-8)
        """
        self.encoding = encoding
    
    def read_lines(self, path: str) -> list[str]:
        """
        Read and clean lines from a text file.
        
        Args:
            path: Path to input file
            
        Returns:
            List of non-empty, stripped lines
            
        Raises:
            FileOperationException: If file cannot be read
        """
        file_path = Path(path)
        
        # Validate file exists
        if not file_path.exists():
            raise FileOperationException(f"File not found: {path}")
        
        # Validate it's a file
        if not file_path.is_file():
            raise FileOperationException(f"Not a file: {path}")
        
        # Read and clean lines
        try:
            with file_path.open("r", encoding=self.encoding) as f:
                lines = [line.strip() for line in f if line.strip()]
            
            if not lines:
                raise FileOperationException(f"File is empty: {path}")
            
            return lines
            
        except UnicodeDecodeError as e:
            raise FileOperationException(
                f"Encoding error in {path}: {e}. Expected {self.encoding}."
            )
        except Exception as e:
            raise FileOperationException(f"Error reading {path}: {e}")
    
    def validate_file(self, path: str) -> bool:
        """
        Validate that a file can be read.
        
        Args:
            path: Path to validate
            
        Returns:
            True if file is valid and readable
        """
        try:
            self.read_lines(path)
            return True
        except FileOperationException:
            return False

File: src/rust-app/src/infra/file_reader.rs
//! File reading infrastructure. Ports `src/infrastructure/file_reader.py`.

use std::fs;
use std::path::Path;

use crate::domain::errors::{AppError, Result};

/// Reads text files containing network records, one per line.
pub struct TextFileReader {
    /// Declared text encoding. Only UTF-8 is supported (the application default).
    pub encoding: String,
}

impl TextFileReader {
    pub fn new(encoding: impl Into<String>) -> Self {
        Self {
            encoding: encoding.into(),
        }
    }

    /// Read and clean lines from a text file.
    ///
    /// Returns the non-empty, trimmed lines. Each line is trimmed of
    /// surrounding whitespace and blank lines are dropped, matching the
    /// Python `[line.strip() for line in f if line.strip()]`.
    pub fn read_lines(&self, path: &str) -> Result<Vec<String>> {
        let file_path = Path::new(path);

        if !file_path.exists() {
            return Err(AppError::File(format!("File not found: {path}")));
        }
        if !file_path.is_file() {
            return Err(AppError::File(format!("Not a file: {path}")));
        }

        let bytes = fs::read(file_path)
            .map_err(|e| AppError::File(format!("Error reading {path}: {e}")))?;

        let content = String::from_utf8(bytes).map_err(|_| {
            AppError::File(format!(
                "Encoding error in {path}: invalid byte sequence. Expected {}.",
                self.encoding
            ))
        })?;

        let lines: Vec<String> = content
            .split('\n')
            .map(|line| line.trim().to_string())
            .filter(|line| !line.is_empty())
            .collect();

        if lines.is_empty() {
            return Err(AppError::File(format!("File is empty: {path}")));
        }

        Ok(lines)
    }

    /// Validate that a file can be read.
    pub fn validate_file(&self, path: &str) -> bool {
        self.read_lines(path).is_ok()
    }
}
